swarm/src/lib.rs

```diff
@@ -415,27 +415,65 @@ where
 
         let peer_id = dial_opts.get_peer_id();
         let condition = dial_opts.peer_condition();
+        let dial_conditions = dial_opts.dial_conditions();
         let connection_id = dial_opts.connection_id();
 
-        let should_dial = match (condition, peer_id) {
-            (PeerCondition::Always, _) => true,
-            (PeerCondition::Disconnected, None) => true,
-            (PeerCondition::NotDialing, None) => true,
-            (PeerCondition::Disconnected, Some(peer_id)) => !self.pool.is_connected(peer_id),
-            (PeerCondition::NotDialing, Some(peer_id)) => !self.pool.is_dialing(peer_id),
-        };
+        // `peer_condition` is deprecated; only use if explicitly set (`Some`).
+        // Block to be removed once `peer_condition` is removed.
+        if let Some(condition) = condition {
+            let false_condition = match (condition, peer_id) {
+                (PeerCondition::Disconnected, Some(peer_id))
+                    if !self.pool.is_connected(peer_id) =>
+                {
+                    Some(PeerCondition::Disconnected)
+                }
+                (PeerCondition::NotDialing, Some(peer_id)) if !self.pool.is_dialing(peer_id) => {
+                    Some(PeerCondition::NotDialing)
+                }
+                _ => None,
+            };
 
-        if !should_dial {
-            let e = DialError::DialPeerConditionFalse(condition);
+            if let Some(condition) = false_condition {
+                let e = DialError::DialPeerConditionFalse(condition);
 
-            self.behaviour
-                .on_swarm_event(FromSwarm::DialFailure(DialFailure {
-                    peer_id,
-                    error: &e,
-                    connection_id,
-                }));
+                self.behaviour
+                    .on_swarm_event(FromSwarm::DialFailure(DialFailure {
+                        peer_id,
+                        error: &e,
+                        connection_id,
+                    }));
 
-            return Err(e);
+                return Err(e);
+            }
+        } else {
```

Comment:
This `else` causes us to only evaluate one or the other right?

I think that is not what we want because it essentially disables the new `DialConditions` if one is still using the old `PeerCondition`. That is a rather subtle detail.

What speaks against always evaluating both? If one specifies a `PeerCondition` that results in the dial going ahead but a `DialCondition` that would prevent it, shouldn't the dial be prevented?

Reply:
This is a big misunderstanding on my part then. I was assuming that we do not want mix both the deprecated and new behaviors. Evaluating both makes sense I guess. If someone is using the deprecated conditions, then they'll probably not be using the new conditions.